### PR TITLE
buffer: replace generic error with Box<Error>

### PR DESCRIPTION
**Pull request by carllerche**
Work towards #131 

Refs: #156

### CODE REVIEW

File: tower-buffer/tests/buffer.rs [hunk 1]
```diff
@@ -83,29 +84,55 @@ fn when_inner_is_not_ready() {
...
     let mut res1 = service.call("hello");
 
     // Allow the Buffer's executor to do work
     ::std::thread::sleep(::std::time::Duration::from_millis(100));
     with_task(|| {
         let e = res1.poll().unwrap_err();
-        if let error::Error::Closed(e) = e {
-            assert!(format!("{:?}", e).contains("poll_ready"));
-            assert_eq!(e.error(), &tower_mock::Error::Other("foobar"));
+        if let Some(e) = e.downcast_ref::<error::ServiceError>() {
```
Comment (carllerche) [inline tower-buffer/tests/buffer.rs]: Most of this junk will go away once `tower-mock` is updated to use `Box<Error>`.

File: tower-mock/src/lib.rs [hunk 2]
```diff
@@ -336,6 +337,32 @@ impl<T, E> Future for ResponseFuture<T, E> {
     }
 }
 
+// ===== impl Error =====
+
+impl<T> fmt::Display for Error<T>
```
Comment (carllerche) [inline tower-mock/src/lib.rs]: This change is just to get things to work for now. It will be changed once `tower-mock` is updated to use `Box<Error>`.

**Comment by carllerche** [comment]
This breaks a bunch of the other tower crates... I will be working on that soon.

File: tower-balance/examples/demo.rs [hunk 3]
```diff
@@ -233,12 +234,25 @@ impl Discover for Disco {
     }
 }
 
+#[derive(Debug)]
+pub struct DiscoError;
```
Comment (carllerche) [inline tower-balance/examples/demo.rs]: This is all pretty terrible, but I *think* most of these changes will go away *once* `Box<Error>` is applied everywhere.

File: tower-balance/examples/demo.rs [hunk 4]
```diff
@@ -305,3 +323,31 @@ where
         Ok(Async::Ready(None))
     }
 }
+
+pub struct Errify<T>(T);
```
Comment (LucioFranco) [inline tower-balance/examples/demo.rs]: is this something we will always need? or is it just for compat between other services that do not yet use `Box<Error>`?

Comment (carllerche) [reply tower-balance/examples/demo.rs]: It should go away once everything is `Box<Error>`.

File: tower-buffer/src/error.rs [hunk 5]
```diff
@@ -5,9 +5,9 @@ use std::sync::Arc;
 
 /// An error produced by a `Service` wrapped by a `Buffer`
 #[derive(Debug)]
-pub struct ServiceError<E> {
+pub struct ServiceError {
```
Comment (olix0r) [inline tower-buffer/src/error.rs]: should this impl clone now that it can?

Comment (carllerche) [reply tower-buffer/src/error.rs]: Generally, errors don't implement `Clone`, so i thought i would avoid it unless needed.

**Pull request merged**